sympy/geometry/ellipse.py

```diff
@@ -171,13 +173,30 @@ def _do_ellipse_intersection(self, o):
         Private helper method for `intersection`.
 
         """
+        from sympy.geometry.polygon import Polygon
+        # do a quick check
+        def box(o):
+            x1, y1, x2, y2 = o.bounds
+            return Polygon((x1, y1), (x2, y1), (x2, y2), (x1, y2))
+        try:
+            if not box(o).intersection(box(self)):
+                return []
+        except Undecidable:
+            pass
+        # there is a potential intersection so proceed
         x = Dummy('x', real=True)
         y = Dummy('y', real=True)
         seq = self.equation(x, y)
         oeq = o.equation(x, y)
-        # TODO: Replace solve with nonlinsolve, when nonlinsolve will be able to solve in real domain
-        result = solve([seq, oeq], x, y)
-        return [Point(*r) for r in list(uniq(result))]
+        result = nonlinsolve([seq, oeq], (x, y))
+        ok = []
+        for r in list(uniq(result)):
+            try:
+                r = Point([sqrtdenest(i) for i in r])
+            except ValueError:
+                continue  # there was an imaginary coord: invalid soln
+            ok.append(r)
+        return list(ordered(ok))
```

Comment:
If you asked me, I would wait with that change, when `nonlinsove_real` will be ready. Your proposition is good and correct, but filtering should be done in `nonlinsove`. Now it complicated a bit readability of code.